mage_ai/data_preparation/models/pipeline.py

```diff
@@ -58,6 +58,35 @@ def create(self, name, repo_path):
             yaml.dump(dict(name=name, uuid=uuid), fp)
         return Pipeline(uuid, repo_path)
 
+    @classmethod
+    def duplicate(cls, source_pipeline: "Pipeline", duplicate_pipeline_name: str):
+        duplicate_pipeline = cls.create(duplicate_pipeline_name, source_pipeline.repo_path)
+        # first pass to load blocks
+        for block_uuid in source_pipeline.blocks_by_uuid:
+            source_block = source_pipeline.blocks_by_uuid[block_uuid]
+            new_block = Block.get_block(source_block.name, source_block.uuid, source_block.type)
+            duplicate_pipeline.add_block(new_block)
+        # second pass to make connections
+        for block_uuid in source_pipeline.blocks_by_uuid:
+            source_block = source_pipeline.blocks_by_uuid[block_uuid]
+            duplicate_block = duplicate_pipeline.blocks_by_uuid[block_uuid]
+            duplicate_block.upstream_blocks = source_block.upstream_blocks
```

Comment:
let's map it to the block in duplicate_pipeline

Reply:
Changed to now reference blocks from duplicate pipelines.